src/states/proving_node.rs

```diff
@@ -0,0 +1,601 @@
+        our_section: (Prefix<XorName>, BTreeSet<PublicId>),
+        action_sender: RoutingActionSender,
+        cache: Box<Cache>,
+        crust_service: Service,
+        old_full_id: FullId,
+        new_full_id: FullId,
+        min_section_size: usize,
+        proxy_pub_id: PublicId,
+        timer: Timer,
+    ) -> Self {
+        let dev_config = config_handler::get_config().dev.unwrap_or_default();
+        let public_id = *new_full_id.public_id();
+
+        let mut peer_mgr = PeerManager::new(public_id, dev_config.disable_client_rate_limiter);
+        peer_mgr.insert_peer(Peer::new(proxy_pub_id, PeerState::Proxy));
+
+        let challenger_count = our_section.1.len();
+        let resource_prover = ResourceProver::new(action_sender, timer.clone(), challenger_count);
+
+        let mut node = Self {
+            crust_service,
+            ack_mgr: AckManager::new(),
+            old_full_id,
+            full_id: new_full_id,
+            msg_queue: VecDeque::new(),
+            msg_backlog: Vec::new(),
+            min_section_size,
+            peer_mgr,
+            cache,
+            routing_msg_filter: RoutingMessageFilter::new(),
+            timer,
+            resource_prover,
+            disable_resource_proof: dev_config.disable_resource_proof,
+            joining_prefix: our_section.0,
+            notified_nodes: Default::default(),
+        };
+        node.start(our_section.1, &proxy_pub_id);
+        node
+    }
+
+    /// Called immediately after construction. Sends `ConnectionInfoRequest`s to all members of
+    /// `our_section` to then start the candidate approval process.
+    fn start(&mut self, our_section: BTreeSet<PublicId>, proxy_pub_id: &PublicId) {
+        self.resource_prover.start(self.disable_resource_proof);
+
+        trace!("{} Relocation completed.", self);
+        info!(
+            "{} Received relocation section. Establishing connections to {} peers.",
+            self,
+            our_section.len()
+        );
+
+        let src = Authority::Client {
+            client_id: *self.full_id.public_id(),
+            proxy_node_name: *proxy_pub_id.name(),
+        };
+        // There will be no events raised as a result of these calls, so safe to just use a
+        // throwaway `EventBox` here.
+        let mut outbox = EventBuf::new();
+        for pub_id in &our_section {
+            debug!(
+                "{} Sending connection info request to {:?} on Relocation response.",
+                self, pub_id
+            );
+            let dst = Authority::ManagedNode(*pub_id.name());
+            if let Err(error) = self.send_connection_info_request(*pub_id, src, dst, &mut outbox) {
+                debug!(
+                    "{} - Failed to send connection info request to {:?}: {:?}",
+                    self, pub_id, error
+                );
+            }
+        }
+    }
+
+    pub fn handle_action(&mut self, action: Action, outbox: &mut EventBox) -> Transition {
```

Comment:
With the current pattern these seem to be multiple somewhat copy pasted implementation.

Reply:
There are differences though. Don't see an obvious way to deduplicate it without making the code convoluted. Might be worth to think about more radical reorganization of the code to avoid this.